text/0192-remove-constructs-compat.md

```diff
@@ -0,0 +1,667 @@
+
+[monolithic packaging]: https://github.com/aws/aws-cdk-rfcs/blob/master/text/0006-monolothic-packaging.md
+
+# Design
+
+This section analysis the required changes and discusses the implementation
+approach and alternatives.
+
+This design is based on this [proof of concept](https://github.com/aws/aws-cdk/pull/8962).
+
+## Removal of the base types (`cdk.Construct`, `cdk.IConstruct`, ...)
+
+Once `construct-compat.ts` is removed from `@aws-cdk/core`, all CDK code (both
+the framework itself and user code) would need to be changed to use the types
+from `constructs`.
+
+Since the APIs are similar in almost all cases, this is a simple mechanical
+change as shown in the [release notes](#release-notes).
+
+The main concern for the CDK codebase is maintaining this change alongside a 1.x
+branch until we switch over to 2.x. Since this change includes modifications to
+all `import` sections in almost all of our files, merge conflicts are imminent.
+
+Possible solutions:
+1. Take the hit and resolve the conflicts manually every time. These are
+   mechanical fixes and easy to do.
+2. Add some automation using eslint to fix these merge conflicts.
+3. Perform the change on the 1.x codebase such that the soon-to-be-moved types
+   are imported separately from other types of `@aws-cdk/core`. This will
+   violate an eslint rule, but we can disable it for the time being.
+
+We can start with #1 for a while and see how bad it is. If it is, we can try to
+introduce some automation or back-porting.
+
+Something that we can definitely do on 1.x is replace all the variants such as
+`core.Construct` and `cdk.Construct` to just `Construct`. This will slightly
+help.
+
+
+## Removal of "synthesize"
+
+Version 4.x of the `constructs` library does not contain a lifecycle hook for
+synthesis.
+
+### Motivation
+
+The reason this is not available at the base class is because the abstraction
+did not "hold water" as the AWS CDK evolved and new CDKs emerged.  In the AWS
+CDK, we eventually ended up with a centralized synthesis logic at the
+`Stage`-level ([TODO: ref]). The main reason was that we needed to "break" the
+recursion in various domain-specific points (e.g. stages, nested stacks) which
+meant that the generic logic of "traverse the entire tree and call `synthesize`"
+did not hold. In `cdk8s`, the support for [TODO: chart dependencies] required
+that the name of the output manifest will be determined based on the topologic
+order at the app level. Here again, the generic approach failed.
+
+In leu of those failures, we decided that there is no additional value in
+actually offering a synthesis mechanism at the `constructs` level. Each
+CDK-domain implements synthesis at the "right" level. This does not mean that
+specific domains can't offer a decentralized approach (i.e. call a method called
+"synthesize" on all constructs in the tree), it just means that this is not
+provided at the base `Construct` class.
```

Comment:
Conceptually okay with this, but I would love `constructs` to provide documentation on how to "properly" to that. In the current state of things, it's a little hand-wavy.

Reply:
I agree that this should be discussed in the docs for `constructs`.